tests/phpunit/CRM/Contribute/Form/ContributionTest.php

```diff
@@ -1324,6 +1324,50 @@ public function testCardTypeAndPanTruncation() {
     $this->assertEquals(CRM_Utils_Array::value('pan_truncation', $financialTrxn), 4567);
   }
 
+  /**
+   * Check payment processor is correctly assigned for a contribution page.
+   */
+  public function testContributionBasePreProcess() {
+    //Create contribution page with only pay later enabled.
+    $params = array(
+      'title' => "Test Contribution Page",
+      'financial_type_id' => 1,
+      'currency' => 'NZD',
+      'goal_amount' => 100,
+      'is_pay_later' => 1,
+      'is_monetary' => TRUE,
+      'is_active' => TRUE,
+      'is_email_receipt' => TRUE,
+      'receipt_from_email' => 'yourconscience@donate.com',
+      'receipt_from_name' => 'Ego Freud',
+    );
+
+    $page1 = $this->callAPISuccess("contribution_page", 'create', $params);
+
+    //Execute CRM_Contribute_Form_ContributionBase preProcess
+    //and check the assignment of payment processors
+    $form = new CRM_Contribute_Form_ContributionBase();
+    $form->controller = new CRM_Core_Controller();
+    $form->set('id', $page1['id']);
+    $form->preProcess();
+    $this->assertEquals($form->_paymentProcessor['name'], 'pay_later');
+
+    //Disable all the payment processor for the contribution page.
+    $params['is_pay_later'] = 0;
+    $page2 = $this->callAPISuccess("contribution_page", 'create', $params);
+
+    //Assert an exception is thrown on loading the contribution page.
+    $form = new CRM_Contribute_Form_ContributionBase();
+    $form->controller = new CRM_Core_Controller();
+    $form->set('id', $page2['id']);
+    try {
+      $form->preProcess();
+    }
+    catch (CRM_Core_Exception $e) {
```

Comment:
If the expectation is that an extension is thrown here, this will still pass if one is not. You can add a return in the catch box & a $this->markFail('exception was expected'); to deal with that

Reply:
Oh Thanks, seems I was extra optimistic here ;-)